### ISSUE TITLE
Useless, incorrect error message when failing to find viable unary `operator*`

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#72217]
Given the following invalid translation unit:

```cpp
struct a {
};
struct b {
};
void operator*(b) {
}

void g() {
	a x;
	*x;
}
```

clang emits the error message

```console
<source>:10:2: error: indirection requires pointer operand ('a' invalid)
   10 |         *x;
      |         ^~
1 error generated.
Compiler returned: 1
```

https://godbolt.org/z/bbvs4qq6n

This doesn't tell me much about what went wrong, and it outright lies (you can say `*x` for types other than pointers). Compare that to the error message if the failing function is anything else:

```cpp
struct a {
};
struct b {
};
void f(b) {
}

void g() {
	a x;
	f(x);
}
```

And clang fails with

```console
<source>:10:2: error: no matching function for call to 'f'
   10 |         f(x);
      |         ^
<source>:5:6: note: candidate function not viable: no known conversion from 'a' to 'b' for 1st argument
    5 | void f(b) {
      |      ^ ~
1 error generated.
Compiler returned: 1
```

https://godbolt.org/z/WTMvWWTro

**Comment by shafik** [comment]
Confirmed.

The diagnostics involved are `err_typecheck_indirection_requires_pointer` in `CheckIndirectionOperand` and `err_member_function_call_bad_cvr`	 in `Sema::PerformImplicitObjectArgumentInitialization`.
